src/multihash.rs

```diff
@@ -85,47 +89,105 @@ pub struct RawMultihash {
     digest: crate::UnknownDigest<crate::U32>,
 }
 
-impl MultihashDigest for RawMultihash {
-    fn code(&self) -> u64 {
+impl RawMultihash {
+    /// Wraps the digest in a multihash.
+    pub fn wrap(code: u64, digest: &[u8]) -> Result<Self, Error> {
+        Ok(Self {
+            code,
+            size: digest.len() as _,
+            digest: Digest::wrap(digest)?,
+        })
+    }
+
+    /// Returns the code of the multihash.
+    pub fn code(&self) -> u64 {
         self.code
     }
 
-    fn size(&self) -> u8 {
+    /// Returns the size of the digest.
+    pub fn size(&self) -> u8 {
         self.size
     }
 
-    fn digest(&self) -> &[u8] {
+    /// Returns the digest.
+    pub fn digest(&self) -> &[u8] {
         self.digest.as_ref()
     }
 
+    /// Reads a multihash from a byte stream.
     #[cfg(feature = "std")]
-    fn read<R: std::io::Read>(r: R) -> Result<Self, Error>
+    pub fn read<R: std::io::Read>(r: R) -> Result<Self, Error>
     where
         Self: Sized,
     {
         let (code, size, digest) = read_multihash(r)?;
         Ok(Self { code, size, digest })
     }
+
+    /// Parses a multihash from a bytes.
+    #[cfg(feature = "std")]
+    pub fn from_bytes(mut bytes: &[u8]) -> Result<Self, Error>
+    where
+        Self: Sized,
+    {
+        Self::read(&mut bytes)
+    }
+
+    /// Writes a multihash to a byte stream.
+    #[cfg(feature = "std")]
+    pub fn write<W: std::io::Write>(&self, w: W) -> Result<(), Error> {
+        write_multihash(w, self.code(), self.size(), self.digest())
+    }
+
+    /// Returns the bytes of a multihash.
+    #[cfg(feature = "std")]
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bytes = vec![];
+        self.write(&mut bytes)
+            .expect("writing to a vec should never fail");
+        bytes
+    }
+
+    /// From Multihash.
+    pub fn from_mh<MH: MultihashDigest>(mh: &MH) -> Result<Self, Error> {
+        use generic_array::GenericArray;
+
+        // raw multihash is backed by a [u8; 32]
+        if mh.size() > 32 {
+            return Err(Error::InvalidSize(mh.size() as _));
+        }
+        let mut digest = GenericArray::default();
+        digest[..mh.size() as _].copy_from_slice(mh.digest());
+        Ok(Self {
+            code: mh.code(),
+            size: mh.size(),
+            digest: digest.into(),
+        })
+    }
+
+    /// To Multihash.
+    pub fn to_mh<MH: MultihashDigest>(&self) -> Result<MH, Error> {
+        MH::wrap(self.code(), self.digest())
+    }
```

Comment:
Could those to implemented with the [`TryFrom` trait](https://doc.rust-lang.org/std/convert/trait.TryFrom.html)? I think that would be cleaner and more versatile.

I'd also do the same for `from_bytes`/`to_bytes` on the `MultihashDigest` trait to keep it small (though that's a separate PR, I'll open an issue about other changes I'd like to see soon).

Reply:
The difference is that to_mh takes a &self instead of try_into(self). The way to use &self references with Into traits is annoying, so you end up having to do either Multihash::try_from(&raw) or (&raw).try_into().

Reply:
I don't see an issue using `Multihash::try_from(&raw)`. What I like about it is, that it is a pattern within the Rust ecosystem that you learn once and then can apply it to other libraries (if they support it) easily.

Though if you want those, we can just implement both things (I'm not a fan of that, but that sounds like a good compromise). The `TryFrom` trait implementation can again be a separate PR.

Reply:
The other issue is that TryFrom doesn't work with a generic parameter, due to there already being an impl in std for all From<T>